### PR TITLE
Bpb 373

### PR DESCRIPTION
**Pull request by scribe**
Isolated the creation of jobs into job factories so we no longer have a ton of different versions of the creation code out there. The presenters now are in charge of mapping UI data into sdk-friendly data, and the factories are in charge of handling the creation of jobs and tasks.

### CODE REVIEW

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/ds3panel/Ds3PanelPresenter.java [hunk 1]
```diff
@@ -691,30 +695,10 @@ private void startGetJob(final List<Ds3TreeTableValueCustom> listFiles,
...
-                        ds3CancelSingleJobTask.setOnFailed(SafeHandler.logHandle(event -> LOG.error("Failed to cancel job")));
-                        ds3CancelSingleJobTask.setOnSucceeded(SafeHandler.logHandle(event -> {
-                            LOG.info("Get Job cancelled");
-                            loggingService.logMessage("GET Job Cancelled", LogType.INFO);
-                            getTreeTableView().refresh();
-                        }));
-                        workers.execute(ds3CancelSingleJobTask);
-                    }));
-                    jobWorkers.execute(jobTask);
+                        return null;
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/ds3panel/Ds3PanelPresenter.java]: Why is this returning null?

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/interruptedjobwindow/JobInfoPresenter.java [hunk 2]
```diff
@@ -243,39 +248,11 @@ protected Object call() {
...
-                    ds3CancelSingleJobTask.setOnFailed(SafeHandler.logHandle((WorkerStateEvent event) -> {
-                        LOG.error("Cancellation of recovered job failed", event.getSource().getException());
-                        loggingService.logMessage("Cancellation of recoved job failed", LogType.ERROR);
-                    }));
-
-                    workers.execute(ds3CancelSingleJobTask);
-                }));
-
-                jobWorkers.execute(jobTask);
+                    return null;
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/interruptedjobwindow/JobInfoPresenter.java]: Why is this returning null?

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/interruptedjobwindow/JobInfoPresenter.java [hunk 3]
```diff
@@ -312,34 +289,11 @@ public void saveJobFileDialog() {
...
-                        LOG.error("Failed to cancedl job", e);
-                    } finally {
-                        final Map<String, FilesAndFolderMap> jobIDMapSecond = ParseJobInterruptionMap.removeJobID(jobInterruptionStore, key, endpointInfo.getEndpoint(), deepStorageBrowserPresenter, loggingService);
-                        ParseJobInterruptionMap.setButtonAndCountNumber(jobIDMapSecond, deepStorageBrowserPresenter);
-                        RefreshCompleteViewWorker.refreshCompleteTreeTableView(ds3Common, workers, dateTimeUtils, loggingService);
-                    }
-                }));
-
-                jobWorkers.execute(recoverInterruptedJob);
+                    return null;
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/interruptedjobwindow/JobInfoPresenter.java]: Same question about null

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/localfiletreetable/LocalFileTreeTablePresenter.java [hunk 4]
```diff
@@ -487,73 +496,20 @@ private void startGetJob(final List<Ds3TreeTableValueCustom> listFiles,
...
-                        ds3CancelSingleJobTask.setOnSucceeded(SafeHandler.logHandle(event -> {
-                            LOG.info("Get Job canceled");
-                            loggingService.logMessage("GET Job Canceled", LogType.INFO);
-                            refreshFileTreeView();
-                        }));
-                        workers.execute(ds3CancelSingleJobTask);
-                    }));
-                    jobWorkers.execute(jobTask);
+                            .map(ds3 -> new kotlin.Pair<>( ds3.getFullName(), ds3.getParent())) .collect(GuavaCollectors.immutableList());
+                    getJobFactory.create(fileAndParent, bucket, localPath, client, () -> { refreshFileTreeView(); return null; });
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/localfiletreetable/LocalFileTreeTablePresenter.java]: Not the nulls!

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/localfiletreetable/LocalFileTreeTablePresenter.java [hunk 5]
```diff
@@ -487,73 +496,20 @@ private void startGetJob(final List<Ds3TreeTableValueCustom> listFiles,
...
-            LOG.info("BULK_PUT job {} Canceled.", jobId);
-            loggingService.logMessage(resourceBundle.getString("putJobCancelled"), LogType.SUCCESS);
-
-            ParseJobInterruptionMap.removeJobID(jobInterruptionStore, jobId.toString(),
-                    client.getConnectionDetails().getEndpoint(), deepStorageBrowserPresenter, loggingService);
-
-            refreshBlackPearlSideItem(remoteDestination);
-        }));
-        jobWorkers.execute(jobTask);
+            return null;
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/localfiletreetable/LocalFileTreeTablePresenter.java]: You guessed it

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/factories/FactoryExtensions.kt [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    jobId.exists {
+        try {
+            Platform.runLater {
+                client.cancelJobSpectraS3(CancelJobSpectraS3Request(it))
+            }
+        } catch (e: IOException) {
+            log.error("Failed to cancel $type job", e)
+            loggingService.logMessage("Could not cancel $type job", LogType.ERROR)
+        }
+        log.info("$type Job cancelled")
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/factories/FactoryExtensions.kt]: This should use the logging format:

```kotlin
log.info("{} job cancelled", type)
```

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/factories/FactoryExtensions.kt [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+                     type: String): (WorkerStateEvent) -> Unit = { worker: WorkerStateEvent ->
+    val throwable: Throwable = worker.source.exception
+    jobId.exists {
+        removeJob(it, client, jobInterruptionStore, deepStorageBrowserPresenter, loggingService)
+    }
+    log.error("$type Job failed", throwable)
+    loggingService.logMessage("$type Job failed with message: ${throwable.message}", LogType.ERROR)
+}
+
+fun JobTask.onCancelled(client: Ds3Client,
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/factories/FactoryExtensions.kt]: Rather than being an extension method, should this be a method on jobTask?  I ask because of all the dependencies this method requires.

File: dsb-util/src/main/java/com/spectralogic/dsbrowser/util/FunctionalExtensions.kt [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+ *   or in the "license" file accompanying this file.
+ *   This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+ *   CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ *   specific language governing permissions and limitations under the License.
+ * ***************************************************************************
+ */
+
+package com.spectralogic.dsbrowser.util
+
+inline fun <T, R> ((T) -> R).andThen(crossinline block: () -> Unit): (T) -> R {
```
Comment (rpmoore) [inline dsb-util/src/main/java/com/spectralogic/dsbrowser/util/FunctionalExtensions.kt]: You should add some javadoc to explain what this method does, specifically the inlining behavior of it as well.

**Comment by rpmoore** [comment]
Out of date.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/ds3panel/ds3treetable/Ds3TreeTablePresenter.java [hunk 9]
```diff
@@ -696,39 +702,14 @@ private void handleDragDetectedEvent(final Event event) {
     private void startPutJob(final Ds3Client client,
             final List<Pair<String, Path>> files,
             final String bucket,
-            final String targetDir,
-            final JobInterruptionStore jobInterruptionStore) {
-
-        final ImmutableList.Builder<kotlin.Pair<String, Path>> builder = ImmutableList.builder();
+            final String targetDir) {
+        final ImmutableList.Builder<kotlin.Pair<String,Path>> builder = ImmutableList.builder();
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/ds3panel/ds3treetable/Ds3TreeTablePresenter.java]: This seems like it can be replaced with a stream call on files with a collect.  Thoughts?

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/interruptedjobwindow/JobInfoPresenter.java [hunk 10]
```diff
@@ -88,31 +91,34 @@
...
-                            final JobTaskElement jobTaskElement,
-                            final SavedJobPrioritiesStore savedJobPrioritiesStore,
-                            final DeepStorageBrowserPresenter deepStorageBrowserPresenter) {
+            final Ds3Common ds3Common,
+            final Workers workers,
+            final JobWorkers jobWorkers,
+            final JobInterruptionStore jobInterruptionStore,
+            final ButtonCell.ButtonCellFactory buttonCellFactory,
+            final LoggingService loggingService,
+            final DateTimeUtils dateTimeUtils,
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/interruptedjobwindow/JobInfoPresenter.java]: Does DateTimeUtils need to be injected?  This seems like it should be a static class.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/JobTask.kt [hunk 11]
```diff
@@ -14,19 +14,31 @@
...
+import org.slf4j.Logger
+import java.io.IOException
 import java.util.*
 
 class JobTask(private val wrappedJob: JobFacade) : Ds3JobTask() {
     @Throws(Throwable::class)
     override fun executeJob() {
         ds3Client = wrappedJob.getDs3Client()
-        var throwable : Throwable? = null
+        var throwable: Throwable? = null
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/JobTask.kt]: I'm not sure that this logic for capturing a throwable is needed.  If this call is meant to block on the stream, make an explicit blocking call and the blocking will cause the throwable to get bubbled up.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/factories/FactoryExtensions.kt [hunk 12]
```diff
@@ -0,0 +1,35 @@
...
+import com.spectralogic.dsbrowser.gui.services.jobinterruption.JobInterruptionStore
+import com.spectralogic.dsbrowser.gui.util.ParseJobInterruptionMap
+import com.spectralogic.dsbrowser.util.exists
+import javafx.application.Platform
+import javafx.concurrent.WorkerStateEvent
+import org.slf4j.Logger
+import java.io.IOException
+import java.util.*
+
+fun removeJob(uuid: UUID, client: Ds3Client, jobInterruptionStore: JobInterruptionStore, deepStorageBrowserPresenter: DeepStorageBrowserPresenter, loggingService: LoggingService) {
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/factories/FactoryExtensions.kt]: Instead of creating a global function like this, create a method overload on ParseJobInterruptionMap instead.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/JobTask.kt [hunk 13]
```diff
@@ -55,13 +67,55 @@ class JobTask(private val wrappedJob: JobFacade) : Ds3JobTask() {
...
+            try {
+                Platform.runLater {
+                    client.cancelJobSpectraS3(CancelJobSpectraS3Request(it))
+                }
+            } catch (e: IOException) {
+                log.error("Failed to cancel $type job", e)
+                loggingService.logMessage("Could not cancel $type job", LogType.ERROR)
+            }
+            log.info("{} Job cancelled", type)
+            loggingService.logMessage("$type Job Cancelled", LogType.INFO)
```
Comment (shabtaisharon) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/JobTask.kt]: if the job failed to cancel and you got into the catch you are still going to get here and print that the job was canceled, maybe return in the catch or put this logic in the try block?

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/factories/GetJobFactory.kt [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+class GetJobFactory @Inject constructor(private val loggingService: LoggingService,
+                                        private val jobInterruptionStore: JobInterruptionStore,
+                                        private val resourceBundle: ResourceBundle,
+                                        private val deepStorageBrowserPresenter: DeepStorageBrowserPresenter,
+                                        private val jobWorkers: JobWorkers,
+                                        private val jobTaskElementFactory: JobTaskElement.JobTaskElementFactory) {
+    private companion object {
+        private val LOG: Logger = LoggerFactory.getLogger(GetJobFactory::class.java)
+        private const val TYPE = "Get"
```
Comment (rpmoore) [inline dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/factories/GetJobFactory.kt]: Why is this TYPE "Get" but the put job is, "BULK_PUT"?

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/ds3panel/ds3treetable/Ds3TreeTablePresenter.java (hunk 9 above)
Comment (scribe) [reply dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/ds3panel/ds3treetable/Ds3TreeTablePresenter.java]: Yeah, this would be more elegant as a stream, working on it.

File: dsb-gui/src/test/java/com/spectralogic/dsbrowser/gui/services/jobService/JobServiceTest.kt [hunk 15]
```diff
@@ -89,9 +89,9 @@ class JobServiceTest {
     fun totalJobTest() {
         var total = 100.00
         jobService!!.jobSizeObservable().subscribe(Consumer { t: Number -> total = t.toDouble() })
-        assertThat(total).isEqualTo(PERCENT_INITIAL)
+        assertThat(total).isEqualTo(0.0)
```
Comment (shabtaisharon) [inline dsb-gui/src/test/java/com/spectralogic/dsbrowser/gui/services/jobService/JobServiceTest.kt]: why is this changed from a const to a magic number?

File: dsb-gui/src/test/java/com/spectralogic/dsbrowser/gui/services/jobService/JobServiceTest.kt [hunk 16]
```diff
@@ -89,9 +89,9 @@ class JobServiceTest {
...
         jobService!!.finishedCompletable().blockingGet()
-        assertThat(total).isEqualTo(PERCENT_RAN)
+        assertThat(total).isEqualTo(1.0)
```
Comment (shabtaisharon) [inline dsb-gui/src/test/java/com/spectralogic/dsbrowser/gui/services/jobService/JobServiceTest.kt]: why is this changed from a const to a magic number?

File: dsb-gui/src/test/java/com/spectralogic/dsbrowser/gui/services/jobService/JobServiceTest.kt (hunk 15 above)
Comment (rpmoore) [reply dsb-gui/src/test/java/com/spectralogic/dsbrowser/gui/services/jobService/JobServiceTest.kt]: Personally, I think it makes the tests clearer...

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/interruptedjobwindow/JobInfoPresenter.java (hunk 10 above)
Comment (scribe) [reply dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/interruptedjobwindow/JobInfoPresenter.java]: In the future DateTimeUtils could be injected to get the date in a localized format, for now it looks kind of funny.

Comment (rpmoore) [reply dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/interruptedjobwindow/JobInfoPresenter.java]: I guess my question is, why is a Util class being injected?  I thought all our Util classes only had static methods.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/factories/GetJobFactory.kt (hunk 14 above)
Comment (scribe) [reply dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/factories/GetJobFactory.kt]: I copied these from the old values, I'd be very happy to clean them up.

Comment (rpmoore) [reply dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/services/jobService/factories/GetJobFactory.kt]: Ok, pick one and make it consistent.  I don't really have a preference on which way.

File: dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/interruptedjobwindow/JobInfoPresenter.java (hunk 10 above)
Comment (scribe) [reply dsb-gui/src/main/java/com/spectralogic/dsbrowser/gui/components/interruptedjobwindow/JobInfoPresenter.java]: Changing to static methods.
